databricks/koalas/indexing.py

```diff
@@ -400,16 +399,43 @@ def _select_rows(self, rows_sel):
                 # If slice is None - select everything, so nothing to do
                 return None, None
             elif len(self._internal.index_columns) == 1:
+                sdf = self._kdf_or_kser._internal.sdf
+                index_column = self._kdf_or_kser.index.to_series()
+                index_data_type = index_column.spark_type
                 start = rows_sel.start
                 stop = rows_sel.stop
+                order_column = sdf[NATURAL_ORDER_COLUMN_NAME]
+
+                # get natural order from '__natural_order__' from start to stop
+                # to keep natural order.
+                start_and_stop = (
+                    sdf.select(index_column._scol, NATURAL_ORDER_COLUMN_NAME)
+                       .where((index_column._scol == start) | (index_column._scol == stop))
+                       .collect())
+
+                start = [row[1] for row in start_and_stop if row[0] == start]
+                start = start[0] if len(start) > 0 else None
+
+                stop = [row[1] for row in start_and_stop if row[0] == stop]
+                stop = stop[0] if len(stop) > 0 else None
+
+                if isinstance(index_data_type, StringType):
+                    index_data_type = LongType()
+                    if start is None and rows_sel.start is not None:
+                        raise KeyError(rows_sel.start)
+                    if stop is None and rows_sel.stop is not None:
+                        raise KeyError(rows_sel.stop)
+                else:
+                    if start is None and stop is None:
+                        start = rows_sel.start
+                        stop = rows_sel.stop
+                        order_column = index_column._scol
```

Comment:
This might cause unexpected behavior changes between whether the slice hit or not.
E.g.,

```py
>>> kdf
    max_speed  shield
10          1       2
30          4       5
20          7       8
>>> kdf.loc[10:20, 'max_speed']
10    1
30    4
20    7
Name: max_speed, dtype: int64
>>> kdf.loc[1:21, 'max_speed']
10    1
20    7
Name: max_speed, dtype: int64
```

Reply:
Btw, seems like pandas throws exceptions when the specified values don't exist or are duplicated.

```py
>>> pdf
    max_speed  shield
10          1       2
30          4       5
20          7       8
30          9      10

>>> pdf.loc[1:20]
Traceback (most recent call last):
...
KeyError: 1

>>> pdf.loc[10:30]
Traceback (most recent call last):
...
KeyError: 'Cannot get right slice bound for non-unique label: 30'
```

Now that we collect the data, I guess we can check the values and follow such behaviors.

Reply:
@ueshin thank you for confirming! i'll fix them properly soon 😄

Reply:
> Btw, seems like pandas throws exceptions when the specified values don't exist or are duplicated.

you're right, but more strictly say, they throws exceptions when the specified values don't exist or are duplicated, and the index order is not monotonically increasing or decreasing. 

```python
>>> pdf
    max_speed  shield
10          1       2
30          4       5
20          7       8

>>> pdf.loc[1:20]
Traceback (most recent call last):
...
ValueError: index must be monotonic increasing or decreasing
...
Traceback (most recent call last):
...
KeyError: 1
```

i think it seems not related with duplicated

```python
>>> pdf
    max_speed  shield
10          1       2
20          4       5
20          7       8
20          3       2
30          9      10
30          2       6
>>>
>>> pdf.loc[1:20]
    max_speed  shield
10          1       2
20          4       5
20          7       8
20          3       2
```

Reply:
I see. It's mixed of the monotonic-increasing and the mis-hit and the duplication ..

Reply:
Yeah it is a bit complicated :confused: